File: client/scala/libs/spark.ml/src/main/scala-2.11/edu/mit/csail/db/ml/modeldb/client/event/TransformEvent.scala
package edu.mit.csail.db.ml.modeldb.client.event

import com.twitter.util.Await
import edu.mit.csail.db.ml.modeldb.client.{ModelDbSyncer, SyncableDataFrame, SyncableTransformer}
import modeldb.ModelDBService.FutureIface
import modeldb.TransformEventResponse
import org.apache.spark.ml.Transformer
import org.apache.spark.ml.util.MLWritable
import org.apache.spark.sql.DataFrame

/**
  * Event indicating that a Transformer transformed a DataFrame.
  *
  * @param transformer - The transformer doing the transforming.
  * @param inputDataframe - The input to the transformer.
  * @param outputDataframe - The output from the transformer.
  */
case class TransformEvent(transformer: Transformer,
                          inputDataframe: DataFrame,
                          outputDataframe: DataFrame) extends ModelDbEvent {
  /**
    * Create the actual TransformEvent.
    * @param mdbs - The syncer. Used to compute the input and output features of the Transformer.
    * @return The TransformEvent.
    */
  def makeEvent(mdbs: ModelDbSyncer) = modeldb.TransformEvent(
    SyncableDataFrame(inputDataframe),
    SyncableDataFrame(outputDataframe),
    SyncableTransformer(transformer),
    mdbs.featureTracker.getInputCols(transformer),
    mdbs.featureTracker.getOutputCols(transformer),
    experimentRunId = mdbs.experimentRun.id
  )


  /**
    * Update object ID mappings based on response.
    * @param ter - The response.
    * @param mdbs - The syncer. This contains the object ID mappings.
    * @return The syncer.
    */
  def associate(ter: modeldb.TransformEventResponse, mdbs: ModelDbSyncer) = {
    mdbs.associateObjectAndId(inputDataframe, ter.oldDataFrameId)
      .associateObjectAndId(outputDataframe, ter.newDataFrameId)
      .associateObjectAndId(transformer, ter.transformerId)
      .associateObjectAndId(this, ter.eventId)
  }

  /**
    * Store the event and update object ID mappings.
    * @param client - The client that exposes the functions that we
    *               call to store objects in the ModelDB.
    * @param mdbs - The ModelDbSyncer, included so we can update the ID
    *             mappings after syncing.
    */
  override def sync(client: FutureIface, mdbs: Option[ModelDbSyncer]): Unit = {
    val res = Await.result(client.storeTransformEvent(makeEvent(mdbs.get)))
    associate(res, mdbs.get)
  }
}

File: client/python/modeldb/events/TransformEvent.py
"""
Event indicating a Transformer transformed a dataframe.
"""
from modeldb.events.Event import Event
from ..thrift.modeldb import ttypes as modeldb_types


class TransformEvent(Event):
    """
    Class for creating and storing TransformEvents
    """

    def __init__(self, old_df, new_df, transformer):
        self.old_df = old_df
        self.new_df = new_df
        self.transformer = transformer

    def make_event(self, syncer):
        """
        Constructs a thrift TransformEvent object with appropriate fields.
        """
        syncable_transformer = syncer.convert_model_to_thrift(self.transformer)
        syncable_dataframe_old = syncer.convert_df_to_thrift(self.old_df)
        syncable_dataframe_new = syncer.convert_df_to_thrift(self.new_df)
        te = modeldb_types.TransformEvent(
            syncable_dataframe_old, syncable_dataframe_new,
            syncable_transformer, [], [],
            syncer.experiment_run.id)
        return te

    def associate(self, res, syncer):
        """
        Stores the server response ids into dictionary.
        """
        syncer.store_object(self.old_df, res.oldDataFrameId)
        syncer.store_object(self.new_df, res.newDataFrameId)
        syncer.store_object(self.transformer, res.transformerId)
        syncer.store_object(self, res.eventId)

    def sync(self, syncer):
        """
        Stores TransformEvent on the server.
        """
        te = self.make_event(syncer)
        thrift_client = syncer.client
        res = thrift_client.storeTransformEvent(te)
        self.associate(res, syncer)
